### PR TITLE
[BIOMAGE-1219] - Navigation checks

### PR DESCRIPTION
**Pull request by aerlaut**
# Background
#### Link to issue 

https://biomage.atlassian.net/browse/BIOMAGE-1219

#### Link to staging deployment URL 

https://ui-agi-nav-check-490.scp-staging.biomage.net/

#### Links to any Pull Requests related to this

N.A.

#### Anything else the reviewers should know about the changes here

- Change mechanism to use Redux because we need to fire events from components coming from different levels of nesting.

# Changes
### Code changes

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

### Testing
- [ ] Unit tests written
- [X] Tested locally with Inframock (with latest production data downloaded with biomage experiment pull)
- [X] Deployed to staging

To set up easy local testing with inframock, follow the instructions here: https://github.com/biomage-ltd/inframock
To deploy to the staging environment, follow the instructions here: https://github.com/biomage-ltd/biomage-utils

### Documentation updates
Is all relevant documentation updated to reflect the proposed changes in this PR?

- [ ] Relevant Github READMEs updated
- [ ] Relevant wiki pages created/updated

### Approvers
- [ ] Approved by a member of the core engineering team
- [ ] (UX changes) Approved by vickymorrison (this is her username, tag her if you need approval)

### Just before merging:
- [ ] After the PR is approved, the `unstage` script in here: https://github.com/biomage-ltd/biomage-utils is executed. This script cleans up your deployment to staging

### Optional
- [ ] Photo of a cute animal attached to this PR

### CODE REVIEW

**Review by vickymorrison — APPROVED** [review]

File: src/components/ChangesNotAppliedModal.jsx [hunk 1]
```diff
@@ -1,37 +1,51 @@
...
       title='Changes not applied'
-      onCancel={onCancel}
+      onCancel={() => dispatch(navigateFromProcessingTo(''))}
       footer={(
         <Space size='large' style={{ display: 'flex', justifyContent: 'center' }}>
           <Button
             type='primary'
             key='run'
-            onClick={onRun}
+            onClick={() => {
```
Comment (cosa65) [inline src/components/ChangesNotAppliedModal.jsx:33]: Why was this moved into the modal? FilesUploadModal doesn't trigger the upload for example

File: src/components/ChangesNotAppliedModal.jsx [hunk 2]
```diff
@@ -1,37 +1,51 @@
...
+            }}
             style={{ width: '100px' }}
           >
             Run
           </Button>
           <Button
             type='primary'
             key='discard'
-            onClick={onDiscard}
+            onClick={() => {
```
Comment (cosa65) [inline src/components/ChangesNotAppliedModal.jsx:44]: Same with this, was was it moved into the modal?

File: src/components/ChangesNotAppliedModal.jsx [hunk 3]
```diff
@@ -1,37 +1,51 @@
...
+
+  const router = useRouter();
+  const dispatch = useDispatch();
+
+  const { changedQCFilters, navigationPath } = useSelector((state) => state.experimentSettings.processing.meta);
 
   return (
     <Modal
-      visible={visible}
+      visible={navigationPath}
```
Comment (cosa65) [inline src/components/ChangesNotAppliedModal.jsx:25]: Why navigation path is being used in visible?

File: src/components/ChangesNotAppliedModal.jsx [hunk 4]
```diff
@@ -1,37 +1,51 @@
...
 const ChangesNotAppliedModal = (props) => {
-  const {
-    steps, visible, onRun, onDiscard, onCancel,
-  } = props;
+  const { experimentId } = props;
+
+  const router = useRouter();
+  const dispatch = useDispatch();
+
+  const { changedQCFilters, navigationPath } = useSelector((state) => state.experimentSettings.processing.meta);
```
Comment (cosa65) [inline src/components/ChangesNotAppliedModal.jsx:21]: Why do we store navigationPath in redux?

File: src/redux/actions/experimentSettings/navigateFromProcessingTo.js [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import {
+  EXPERIMENT_SETTINGS_NAVIGATE_TO,
```
Comment (cosa65) [inline src/redux/actions/experimentSettings/navigateFromProcessingTo.js:2]: I don't feel like this fits in experiment settings

File: src/components/ChangesNotAppliedModal.jsx (hunk 4 above)
Comment (aerlaut) [reply src/components/ChangesNotAppliedModal.jsx:21]: Because it has to be shared between multiple components on different hierarchy levels. The link in ContentWrapper needs it, and the "finish QC" button inside the Data Processing component also needs to know about it.

File: src/components/ChangesNotAppliedModal.jsx (hunk 3 above)
Comment (aerlaut) [reply src/components/ChangesNotAppliedModal.jsx:25]: It's a shortcut basically, if it's an empty string the modal doesn't show.

File: src/components/ChangesNotAppliedModal.jsx (hunk 1 above)
Comment (aerlaut) [reply src/components/ChangesNotAppliedModal.jsx:33]: This modal directly listens to the `navigationPath` to get rendered or not. This is so because this modal needs to be called from both the navigation links and also the finish qc button. Because it listens directly to the store, I thought it would be ok for its actions to also be directly related to the store instead of taking functions from its parent component.

File: src/components/ChangesNotAppliedModal.jsx (hunk 2 above)
Comment (aerlaut) [reply src/components/ChangesNotAppliedModal.jsx:44]: Same as above

File: src/redux/actions/experimentSettings/navigateFromProcessingTo.js (hunk 5 above)
Comment (aerlaut) [reply src/redux/actions/experimentSettings/navigateFromProcessingTo.js:2]: I put it in `experimentSettings` because it's the closest to `changedQCFilters`. I'm not sure where else to put it,

File: src/components/ChangesNotAppliedModal.jsx (hunk 1 above)
Comment (cosa65) [reply src/components/ChangesNotAppliedModal.jsx:33]: The thing is that no other modal is doing this, all of them work through callbacks

File: src/components/ChangesNotAppliedModal.jsx (hunk 3 above)
Comment (cosa65) [reply src/components/ChangesNotAppliedModal.jsx:25]: We should probably use null instead of an empty string, and perhaps to make it more explicit use `navigationPath === null`

File: src/components/ChangesNotAppliedModal.jsx (hunk 1 above)
Comment (cosa65) [reply src/components/ChangesNotAppliedModal.jsx:33]: Perhaps could we wrap this `ChangesNotAppliedModal` in a `NavigationHandler` or something with a better name that handles the navigation requests.

When something dispatches to navigate somewhere it should check, if we currently are in data processing and are trying to go somewhere else with unsaved changes then show modal and handle the callbacks.

Comment (cosa65) [reply src/components/ChangesNotAppliedModal.jsx:33]: and move the callbacks back outside, with this navigation component handling those callbacks via dispatches.

File: src/redux/actions/experimentSettings/navigateFromProcessingTo.js (hunk 5 above)
Comment (cosa65) [reply src/redux/actions/experimentSettings/navigateFromProcessingTo.js:2]: I think it fits at the root level of the store, with a new reducer and all that

Comment (cosa65) [reply src/redux/actions/experimentSettings/navigateFromProcessingTo.js:2]: it's about navigation, so not really linked with any of the already existing aprts of the store.

File: src/components/ChangesNotAppliedModal.jsx (hunk 1 above)
Comment (aerlaut) [reply src/components/ChangesNotAppliedModal.jsx:33]: Hmm... are you suggesting something like a middleware?

Comment (aerlaut) [reply src/components/ChangesNotAppliedModal.jsx:33]: Does that mean the other navigation will also route throught this middleware component?

Comment (cosa65) [reply src/components/ChangesNotAppliedModal.jsx:33]: something like our notifications manager
